restli-server/src/main/java/com/linkedin/restli/internal/server/RestLiMethodInvoker.java

```diff
@@ -331,6 +338,60 @@ public void invoke(final RestLiRequestData requestData,
     }
   }
 
+  /**
+   * Method is used to validate if the request's batch size is under
+   * the allowed max batch size which is defined in the server resource.
+   *
+   * @throws RestLiServiceException if request's batch size is larger than the allowed max batch size.
+   */
+  private void validateMaxBatchSize(RestLiRequestData requestData, ResourceMethodDescriptor method,
+                                    ServerResourceContext resourceContext) throws RestLiServiceException
+  {
+    MaxBatchSizeSchema maxBatchSizeAnnotation = method.getMaxBatchSize();
+    if (maxBatchSizeAnnotation == null || !maxBatchSizeAnnotation.isValidate())
+    {
+      return;
+    }
+
+    int requestBatchSize = getRequestBatchSize(requestData, method, resourceContext);
+    int maxBatchSize = maxBatchSizeAnnotation.getValue();
+
+    if (requestBatchSize > maxBatchSizeAnnotation.getValue())
+    {
+      throw new RestLiServiceException(HttpStatus.S_400_BAD_REQUEST,
+              String.format("The request batch size: %s is larger than the allowed max batch size: %s for method: %s",
+                      requestBatchSize, maxBatchSize, method.getMethodName()));
+    }
+  }
+
+  private int getRequestBatchSize(RestLiRequestData requestData, ResourceMethodDescriptor method, ServerResourceContext resourceContext)
+  {
+    switch (method.getMethodType())
+    {
+      case BATCH_GET:
+      case BATCH_DELETE:
+        return requestData.getBatchKeys().size();
+      case BATCH_UPDATE:
+      case BATCH_PARTIAL_UPDATE:
+        return requestData.getBatchKeyEntityMap().size();
+      case BATCH_CREATE:
+        return requestData.getBatchEntities().size();
+      case BATCH_FINDER:
+        return getBatchFinderCriteriaNumber(method, resourceContext);
+      default:
+        return 0;
+    }
+  }
+
+  private int getBatchFinderCriteriaNumber(ResourceMethodDescriptor method, ServerResourceContext resourceContext)
+  {
+    List<Parameter<?>> parameterList = method.getParameters();
+    Integer criteriaParamIndex = method.getBatchFinderCriteriaParamIndex();
+    Parameter<?> criteriaParam = parameterList.get(criteriaParamIndex);
+    DataList criteriaList = (DataList) resourceContext.getStructuredParameter(criteriaParam.getName());
```

Comment:
This can be a single entity (not DataList) for protocol 1.0.0. There is a special logic to convert it to *Array type when we build args. It would be cleaner to use the args for the batch finder case.

Reply:
You are talking about this logic https://github.com/linkedin/rest.li/blob/99999589f89a168a0528aabc4d25c797b755b57b/restli-server/src/main/java/com/linkedin/restli/internal/server/methods/arguments/ArgumentBuilder.java#L505, right? However, for batchFInder case it will never fall in to this statement:https://github.com/linkedin/rest.li/blob/99999589f89a168a0528aabc4d25c797b755b57b/restli-server/src/main/java/com/linkedin/restli/internal/server/methods/arguments/ArgumentBuilder.java#L225. Because batchFinder criteria param is required to use array of record template: https://linkedin.github.io/rest.li/batch_finder_resource_method#method-annotation-and-parameters. It will go to buildRegularArgument path https://github.com/linkedin/rest.li/blob/99999589f89a168a0528aabc4d25c797b755b57b/restli-server/src/main/java/com/linkedin/restli/internal/server/methods/arguments/ArgumentBuilder.java#L232, because the type is array.
I added a test which has one criteria, it works for both protocol 1.0 and 2.0. The request was build through rest.li request builder. 
I also tested build request by a url. It works for this one "/batchGreeting?bq=searchGreetings&criteria[0].id=1".
If I changed the request to "/batchGreeting?bq=searchGreetings&criteria.id=1" the ArgumentBuilder throws exception here: https://github.com/linkedin/rest.li/blob/99999589f89a168a0528aabc4d25c797b755b57b/restli-server/src/main/java/com/linkedin/restli/internal/server/methods/arguments/ArgumentBuilder.java#L340, because a datamap cannot cast to a datalist. 
It means if the code reaches the validateMaxBatchSize, it should be a valid case, it is a DataList. 
I could change it to use args, like (RecordTemplate[])(arg[criteriaParamIndex]).length. But I do not see much value it brings.
Let me know what do you think. Thanks.

Reply:
For array of records, we generate FooArray classes that extend from WrappingArrayTemplate, which is a sub class of DataTemplate.

I think your logic works for batchFinder because we parse it always as DataList. 
`paramValue.getClass() != DataList.class` will be false in the first check you linked. 

Why we parse it always as DataList? Could be because of the way the data is represented in the URI (and this is an array of complex types).
`criteria[0].id=1` 
in the URI. See the logic here: https://github.com/linkedin/rest.li/blob/master/restli-common/src/main/java/com/linkedin/restli/internal/common/QueryParamsDataMap.java#L288

So for batch finder, your current logic is fine as criteria should be an array of records.